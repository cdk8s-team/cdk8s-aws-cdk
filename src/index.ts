export * from './adapter';
export * from './chart';
export * from './mappers';