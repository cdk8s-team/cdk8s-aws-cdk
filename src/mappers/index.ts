export * from './base';
export * from './ec2-security-group';
export * from './iam-policy';
export * from './iam-role';
export * from './lambda-function';
export * from './rds-db-instance';
export * from './rds-db-subnet-group';
export * from './s3-bucket';
