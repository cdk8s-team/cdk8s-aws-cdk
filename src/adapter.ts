import * as aws from 'aws-cdk-lib';
import * as k from 'cdk8s';
import * as kplus from 'cdk8s-plus-28';
import { Construct } from 'constructs';
import * as kfieldexports from './imports/fieldexports-services.k8s.aws';
import * as k8s from './imports/k8s';
import * as mappers from './mappers';

export interface AwsCdkAdapterProps extends aws.StackProps {

  readonly chart: k.Chart;
}

export class AwsCdkAdapater extends aws.Stack {

  private readonly chart: k.Chart;

  private readonly _mappers: Record<string, mappers.CloudFormationResourceMapper> = {};

  constructor(scope: Construct, id: string, props: AwsCdkAdapterProps) {
    super(scope, id, props);
    this.chart = props.chart;

    this.registerMapper(new mappers.RdsDBInstanceMapper(this.chart));
    this.registerMapper(new mappers.Ec2SecurityGroupMapper(this.chart));
    this.registerMapper(new mappers.RdsDBSubnetGroup(this.chart));
    this.registerMapper(new mappers.IamPolicyMapper(this.chart));
    this.registerMapper(new mappers.IamRoleMapper(this.chart));
    this.registerMapper(new mappers.LambdaFunctionMapper(this.chart));
    this.registerMapper(new mappers.S3BucketMapper(this.chart));
  }

  /**
   * Register a mapper that can transform a specific CloudFormation resource type.
   */
  public registerMapper(mapper: mappers.CloudFormationResourceMapper) {
    this._mappers[mapper.type] = mapper;
  }

  /**
   * Transform CloudFormation resources to ACK resources.
   */
  public transformResources(cfnResources: aws.CfnResource[]) {

    for (const cfnResource of cfnResources) {
      const logicalId = this.getLogicalId(cfnResource);
      const cfnType = cfnResource.cfnResourceType;

      const mapper = this.tryFindMapper(cfnType);

      if (!mapper) {
        throw new Error(`Unable to transform resource of type '${cfnType}'. Mapper is not registered`);
      }

      const cfnTemplateResources = Object.values((cfnResource._toCloudFormation() as any).Resources) as any[];
      const cfnProperties = this.resolveIntrinsics(logicalId, this.resolve(cfnTemplateResources[0].value.Properties.value), mapper);

      const resource = mapper.map(logicalId, cfnProperties);

      // for each resource we patch its name if needed
      if (!cfnProperties[mapper.nameMapping.cfnProperty]) {
        resource.addJsonPatch(k.JsonPatch.add(mapper.nameMapping.specPath, resource.name));
        resource.addJsonPatch(k.JsonPatch.add('/metadata', { name: resource.name }));
      }

      // for each resource we create a config map that will hold its field exports.
      // i.e runtime attributes. these will later be used to transform CloudFormation
      // attributes.
      const configMap = new kplus.ConfigMap(this.chart, `${logicalId}ConfigMap`);

      for (const exportMapping of mapper.exportMappings) {
        new kfieldexports.FieldExport(this.chart, this.fieldExportId(logicalId, exportMapping.attribute), {
          spec: {
            from: {
              path: this.fieldExportPath(exportMapping.field),
              resource: { group: resource.apiGroup, kind: resource.kind, name: resource.name },
            },
            to: {
              kind: kfieldexports.FieldExportSpecToKind.CONFIGMAP,
              name: configMap.name,
            },
          },
        });
      }
    }
  }

  /**
   * Transform CloudFormation attributes into ACK field exports.
   *
   * Note that this is only supported for container environment variables,
   * as only they allow referencing fields from config maps, which the field exports export to.
   *
   * If an ApiObject contains a CloudFormation attribute in any other definition, it will remain
   * as is, it will remain in-place.
   */
  public transformAttributes(apiObjects: k.ApiObject[]) {

    for (const apiObject of apiObjects) {

      const spec: any = apiObject.toJson().spec;

      switch (apiObject.kind) {
        case 'Deployment':
        case 'StatefulSet':
        case 'DaemonSet':
        case 'Job':
          this.patchContainerEnvironment(apiObject, spec?.template?.spec?.containers ?? [], '/spec/template/spec/containers');
          break;
        case 'Pod':
          this.patchContainerEnvironment(apiObject, spec?.containers ?? [], '/spec/containers');
          break;
        default:
          break;
      }
    }
  }

  private patchContainerEnvironment(resource: k.ApiObject, containers: k8s.Container[], path: string) {

    for (let i = 0; i < containers.length; i++) {

      const container = containers[i];
      const envVars = container.env ?? [];

      for (let j = 0; j < envVars.length; j++) {

        const envVar = envVars[j];

        if (!envVar.value) {
          // we only conver value --> valueFrom
          continue;
        }

        if (!aws.Token.isUnresolved(envVar.value)) {
          // no conversion necessary
          continue;
        }

        const resolved = this.resolve(envVar.value);

        if (typeof(resolved) !== 'object') {
          // not sure what this means...skip in the meantime
          continue;
        }

        const key = Object.keys(resolved)[0];
        if (!key.startsWith('Fn::GetAtt')) {
          // not sure what this means...skip in the meantime
          continue;
        }

        const value = Object.values(resolved)[0] as string[];
        const logicalId = value[0];
        const attribute = value[1];

        // locate the field exports that is mapped to this attribute.
        const fieldExport = this.tryFindFieldExport(logicalId, attribute);
        if (!fieldExport) {
          throw new Error(`Unable to transform attribute '${JSON.stringify(resolved)}' for env variable '${envVar.name}' in resource '${resource.name}': Field export not registered.`);
        }

        const fieldExportSpec: kfieldexports.FieldExportSpec = fieldExport.toJson().spec;
        const namespace = fieldExport.metadata.namespace ?? 'default';

        resource.addJsonPatch(k.JsonPatch.remove(`${path}/${i}/env/${j}`));
        resource.addJsonPatch(k.JsonPatch.add(`${path}/${i}/env/${j}`, {
          name: envVar.name,
          valueFrom: {
            configMapKeyRef: {
              key: `${namespace}.${fieldExport.name}`,
              name: fieldExportSpec.to.name,
            },
          },
        }));
      }
    }
  }

  private resolveIntrinsics(logicalId: string, properties: any, mapper: mappers.CloudFormationResourceMapper) {

    const resolved = JSON.stringify(properties, (key, value) => {

      if (typeof(value) !== 'object') {
        // not an attribute
        return value;
      }

      if (Object.keys(value).length !== 1) {
        // definitely not an intrinsic, which takes the form '{ key: value }'
        return value;
      }

      const attribute = Object.keys(value)[0];
      const refLogicalId = Object.values(value)[0];

      if (attribute.startsWith('Fn::')) {
        // ACK doesn't have support for Fn:: intrinsics at the moment.
        // actually -- thats not entirely true, some resources do have a way to pass
        // values by referencing other resources. For example, you can pass a `vpcRef`
        // to a security group, instead of the actual vpc id. but I have yet to list those out,
        // and understand how they work.
        throw new Error(`Unable to resolve intrinsic function '${JSON.stringify(value)}' in property '${key}' of resource '${logicalId}'`);
      }

      if (attribute !== 'Ref') {
        // probably just a complext value, move on...
        return value;
      }

      // we can resolve 'Ref' because we know all resource names at synth time.
      const reference = this.chart.node.findChild(refLogicalId as string);
      switch (mapper.refMapping) {
        case mappers.CloudFormationMapperRefMapping.NAME:
          return k.ApiObject.of(reference).name;
        default:
          throw new Error(`Unsupported ref mapping: ${mapper.refMapping}`);
      }
    });

    return JSON.parse(resolved);

  }

  private tryFindMapper(cfnType: string): mappers.CloudFormationResourceMapper {
    return this._mappers[cfnType];
  }

  private tryFindFieldExport(logicalId: string, attr: string): kfieldexports.FieldExport {
    return this.chart.node.tryFindChild(this.fieldExportId(logicalId, attr)) as kfieldexports.FieldExport;
  }

  private fieldExportId(logicalId: string, attribute: string) {
    return `${logicalId}${attribute}`;
  }

  private fieldExportPath(field: string) {
    return `.status.${field}`;
  }

}
