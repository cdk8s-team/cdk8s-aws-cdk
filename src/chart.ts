import * as aws from 'aws-cdk-lib';
import * as k from 'cdk8s';
import { Construct } from 'constructs';
import * as adapter from './adapter';
import * as mappers from './mappers';

const STACK_SYMBOL = Symbol.for('@aws-cdk/core.Stack');

export class Chart extends k.Chart {

  protected readonly adapter: adapter.AwsCdkAdapater;

  constructor(scope: Construct, id: string, props: k.ChartProps = { }) {
    super(scope, id, props);
    this.adapter = new adapter.AwsCdkAdapater(this, 'AwsCdkAdapter', { chart: this });
    this._inheritStack();

    // hack that allows us to inject code just before cdk8s synthesis.
    // TODO - figure out a proper hook for synthesis in cdk8s (or maybe constructs?)
    this.node.addValidation({ validate: () => this._validate() });
  }

  private _validate(): string[] {

    const cfnResources = this.node.findAll().filter(n => aws.CfnResource.isCfnResource(n)) as aws.CfnResource[];
    const apiObjects = this.node.findAll().filter(n => n instanceof k.ApiObject) as k.ApiObject[];

    // transform all cfn resources to ack resources
    this.adapter.transformResources(cfnResources);

    // transform cfn attributes to field exports
    this.adapter.transformAttributes(apiObjects);

    // hack, we need to return something...
    return [];
  }

  /**
   * Monkey patches the chart to also inherit from an AWS CDK stack.
   * This makes it so the chart can be passed as a scope of AWS CDK constructs.
   */
  private _inheritStack() {
    Object.defineProperty(this, STACK_SYMBOL, { value: true });
    const stackMethods = Object.getOwnPropertyNames(Object.getPrototypeOf(Object.getPrototypeOf(this.adapter)));
    const adapterProperties = Object.getOwnPropertyNames(this.adapter);
    for (const p of [...stackMethods, ...adapterProperties]) {

      if (['constructor', 'node'].includes(p)) {
        continue;
      };

      const v = (this.adapter as any)[p];

      if (typeof(v) === 'function') {
        // this is...adventurous? yeah...
        const n = new Function('...args', `return this.adapter.${p}(...args);`);
        Object.defineProperty(this, p, { value: n });
      } else {
        Object.defineProperty(this, p, { value: v });
      }
    }
  }

  /**
   * Registers a new mapper.
   */
  protected registerMapper(mapper: mappers.CloudFormationResourceMapper) {
    this.adapter.registerMapper(mapper);
  }

}
