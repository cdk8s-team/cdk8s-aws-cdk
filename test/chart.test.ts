import {
  aws_ec2 as ec2,
  aws_rds as rds,
} from 'aws-cdk-lib';
import * as aws from 'aws-cdk-lib';
import * as k from 'cdk8s';
import * as kplus from 'cdk8s-plus-24';
import { Construct } from 'constructs';
import * as src from '../src';

test('rds-db-instance', () => {

  class RdsDBInstanceChart extends src.Chart {

    constructor(scope: Construct, id: string, props: k.ChartProps = {}) {
      super(scope, id, props);

      const vpc = ec2.Vpc.fromVpcAttributes(this, 'Vpc', {
        availabilityZones: ['us-east-1b', 'us-east-1a'],
        vpcId: 'vpc-03c9c89ead26f6840',
        privateSubnetIds: ['subnet-065321f51cb2ac562', 'subnet-0d72f4230848972e9'],
        publicSubnetIds: ['subnet-0847295da218124fa', 'subnet-033821bf98e0a89bf'],
      });

      const securityGroup = ec2.SecurityGroup.fromSecurityGroupId(this, 'SecurityGroup', 'sg-06a21c2d7f87340df');

      const dbInstance = new rds.DatabaseInstance(this, 'DatabaseInstance', {
        engine: rds.DatabaseInstanceEngine.POSTGRES,
        vpc,
        credentials: {
          username: 'postgres',
          password: aws.SecretValue.unsafePlainText('password'),
        },
        securityGroups: [securityGroup],

      });

      const application = new kplus.Deployment(this, 'Application');

      const container = application.addContainer({ image: 'application', port: 8080 });
      container.env.addVariable('DB_ADDRESS', kplus.EnvValue.fromValue(dbInstance.dbInstanceEndpointAddress));
      container.env.addVariable('DB_PORT', kplus.EnvValue.fromValue(dbInstance.dbInstanceEndpointPort));


    }

  }

  const app = new k.App();
  const chart = new RdsDBInstanceChart(app, 'RdsDBInstanceChart');
  expect(k.Testing.synth(chart)).toMatchSnapshot();

});