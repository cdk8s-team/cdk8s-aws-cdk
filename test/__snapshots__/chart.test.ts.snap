// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rds-db-instance 1`] = `
Array [
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "name": "rdsdbinstancechart-application-c87c5287",
    },
    "spec": Object {
      "minReadySeconds": 0,
      "progressDeadlineSeconds": 600,
      "replicas": 2,
      "selector": Object {
        "matchLabels": Object {
          "cdk8s.io/metadata.addr": "RdsDBInstanceChart-Application-c86ef48c",
        },
      },
      "strategy": Object {
        "rollingUpdate": Object {
          "maxSurge": "25%",
          "maxUnavailable": "25%",
        },
        "type": "RollingUpdate",
      },
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "cdk8s.io/metadata.addr": "RdsDBInstanceChart-Application-c86ef48c",
          },
        },
        "spec": Object {
          "automountServiceAccountToken": false,
          "containers": Array [
            Object {
              "env": Array [
                Object {
                  "name": "DB_ADDRESS",
                  "valueFrom": Object {
                    "configMapKeyRef": Object {
                      "key": "default.rdsdbinstance-databaseinstance24d16791endpoint.address-c8d4b7a6",
                      "name": "rdsdbinstan-databaseinstance24d16791configmap-c8ebba6c",
                    },
                  },
                },
                Object {
                  "name": "DB_PORT",
                  "valueFrom": Object {
                    "configMapKeyRef": Object {
                      "key": "default.rdsdbinstancecha-databaseinstance24d16791endpoint.port-c844db15",
                      "name": "rdsdbinstan-databaseinstance24d16791configmap-c8ebba6c",
                    },
                  },
                },
              ],
              "image": "application",
              "imagePullPolicy": "Always",
              "name": "main",
              "ports": Array [
                Object {
                  "containerPort": 8080,
                },
              ],
              "resources": Object {
                "limits": Object {
                  "cpu": "1500m",
                  "memory": "2048Mi",
                },
                "requests": Object {
                  "cpu": "1000m",
                  "memory": "512Mi",
                },
              },
              "securityContext": Object {
                "allowPrivilegeEscalation": false,
                "privileged": false,
                "readOnlyRootFilesystem": true,
                "runAsNonRoot": true,
              },
              "startupProbe": Object {
                "failureThreshold": 3,
                "tcpSocket": Object {
                  "port": 8080,
                },
              },
            },
          ],
          "dnsPolicy": "ClusterFirst",
          "hostNetwork": false,
          "restartPolicy": "Always",
          "securityContext": Object {
            "fsGroupChangePolicy": "Always",
            "runAsNonRoot": true,
          },
          "setHostnameAsFQDN": false,
        },
      },
    },
  },
  Object {
    "apiVersion": "rds.services.k8s.aws/v1alpha1",
    "kind": "DBSubnetGroup",
    "metadata": Object {
      "name": "rdsdbinstancechart-databaseinstancesubnetgroupa4fff2a4-c86e7648",
    },
    "spec": Object {
      "description": "Subnet group for DatabaseInstance database",
      "name": "rdsdbinstancechart-databaseinstancesubnetgroupa4fff2a4-c86e7648",
      "subnetIDs": Array [
        "subnet-065321f51cb2ac562",
        "subnet-0d72f4230848972e9",
      ],
    },
  },
  Object {
    "apiVersion": "v1",
    "immutable": false,
    "kind": "ConfigMap",
    "metadata": Object {
      "name": "databaseinstancesubnetgroupa4fff2a4configmap-c864c144",
    },
  },
  Object {
    "apiVersion": "v1",
    "immutable": false,
    "kind": "Secret",
    "metadata": Object {
      "name": "rdsdbinstancec-databaseinstance24d16791secret-c815fde5",
    },
    "stringData": Object {
      "password": "password",
    },
  },
  Object {
    "apiVersion": "rds.services.k8s.aws/v1alpha1",
    "kind": "DBInstance",
    "metadata": Object {
      "name": "rdsdbinstancechart-databaseinstance24d16791-c8700e61",
    },
    "spec": Object {
      "allocatedStorage": 100,
      "copyTagsToSnapshot": true,
      "dbInstanceClass": "db.m5.large",
      "dbInstanceIdentifier": "rdsdbinstancechart-databaseinstance24d16791-c8700e61",
      "dbSubnetGroupName": "rdsdbinstancechart-databaseinstancesubnetgroupa4fff2a4-c86e7648",
      "engine": "postgres",
      "masterUserPassword": Object {
        "key": "password",
        "name": "rdsdbinstancec-databaseinstance24d16791secret-c815fde5",
      },
      "masterUsername": "postgres",
      "storageType": "gp2",
      "vpcSecurityGroupIDs": Array [
        "sg-06a21c2d7f87340df",
      ],
    },
  },
  Object {
    "apiVersion": "v1",
    "immutable": false,
    "kind": "ConfigMap",
    "metadata": Object {
      "name": "rdsdbinstan-databaseinstance24d16791configmap-c8ebba6c",
    },
  },
  Object {
    "apiVersion": "services.k8s.aws/v1alpha1",
    "kind": "FieldExport",
    "metadata": Object {
      "name": "rdsdbinstance-databaseinstance24d16791endpoint.address-c8d4b7a6",
    },
    "spec": Object {
      "from": Object {
        "path": ".status.endpoint.address",
        "resource": Object {
          "group": "rds.services.k8s.aws",
          "kind": "DBInstance",
          "name": "rdsdbinstancechart-databaseinstance24d16791-c8700e61",
        },
      },
      "to": Object {
        "kind": "configmap",
        "name": "rdsdbinstan-databaseinstance24d16791configmap-c8ebba6c",
      },
    },
  },
  Object {
    "apiVersion": "services.k8s.aws/v1alpha1",
    "kind": "FieldExport",
    "metadata": Object {
      "name": "rdsdbinstancecha-databaseinstance24d16791endpoint.port-c844db15",
    },
    "spec": Object {
      "from": Object {
        "path": ".status.endpoint.port",
        "resource": Object {
          "group": "rds.services.k8s.aws",
          "kind": "DBInstance",
          "name": "rdsdbinstancechart-databaseinstance24d16791-c8700e61",
        },
      },
      "to": Object {
        "kind": "configmap",
        "name": "rdsdbinstan-databaseinstance24d16791configmap-c8ebba6c",
      },
    },
  },
]
`;
